libsql-wal/src/storage/backend/s3.rs

```diff
@@ -1,126 +1,471 @@
-            for object in result.unwrap().contents() {
-                let key = object.key().unwrap();
-
-                let file_name = key.split("/").last().unwrap();
-
-                let (start_frame_no, end_frame_no) =
-                    super::fs::parse_segment_file_name(&file_name)?;
-
-                if start_frame_no <= frame_no && end_frame_no >= frame_no {
-                    let out = self
-                        .client
-                        .get_object()
-                        .bucket(&self.config.bucket)
-                        .key(key)
-                        .send()
-                        .await
-                        .unwrap();
-
-                    let stream = out.body.into_async_read();
-                    return Ok((file_name.to_string(), Box::pin(stream)));
+        let Some(segment_key) = self.find_segment(config, &folder_key, frame_no).await? else {
+            return Err(Error::FrameNotFound(frame_no));
+        };
+        if segment_key.includes(frame_no) {
+            let (_, index) = tokio::try_join!(
+                self.fetch_segment_data(config, &folder_key, &segment_key, dest_path),
+                self.fetch_segment_index(config, &folder_key, &segment_key),
+            )?;
+
+            Ok(index)
+        } else {
+            todo!("not found");
+        }
+    }
+
+    async fn meta(&self, config: &Self::Config, namespace: NamespaceName) -> Result<super::DbMeta> {
+        // request a key bigger than any other to get the last segment
+        let folder_key = FolderKey {
+            cluster_id: &config.cluster_id,
+            namespace: &namespace,
+        };
+
+        let max_segment_key = self.find_segment(config, &folder_key, u64::MAX).await?;
+
+        Ok(super::DbMeta {
+            max_frame_no: max_segment_key.map(|s| s.end_frame_no).unwrap_or(0),
+        })
+    }
+
+    fn default_config(&self) -> Arc<Self::Config> {
+        self.default_config.clone()
+    }
+}
+
+#[derive(Clone, Copy)]
+enum StreamState {
+    Init,
+    WaitingChunk,
+    Done,
+}
+
+struct FileStreamBody<F> {
+    inner: Arc<F>,
+    current_offset: u64,
+    chunk_size: usize,
+    state: StreamState,
+    fut: ReusableBoxFuture<'static, std::io::Result<Bytes>>,
+}
+
+impl<F> FileStreamBody<F> {
+    fn new(inner: F) -> Self {
+        Self::new_inner(inner.into())
+    }
+
+    fn new_inner(inner: Arc<F>) -> Self {
+        Self {
+            inner,
+            current_offset: 0,
+            chunk_size: 4096,
+            state: StreamState::Init,
+            fut: ReusableBoxFuture::new(std::future::pending()),
+        }
+    }
+
+    fn into_byte_stream(self) -> ByteStream
+    where
+        F: FileExt,
+    {
+        let body = SdkBody::retryable(move || {
+            let s = Self::new_inner(self.inner.clone());
+            SdkBody::from_body_1_x(s)
+        });
+
+        ByteStream::new(body)
+    }
+}
+
+impl<F> http_body::Body for FileStreamBody<F>
```

Comment:
I don't totally follow why we need a custom body implementation? Might be worth adding a comment why here

Reply:
I couldn't find how to wrap a stream, but I reckon we may have used https://docs.rs/stream-body/latest/stream_body/